### PR TITLE
SICP package test

### PR DESCRIPTION
**Pull request by Hungkhoaitay**
Create a Node environment and use virtual machine to test functions in the SICP package.
Testing bases on the old testing of SICP.

### CODE REVIEW

**Comment by samuelfangjw** [comment]
Talked to @Hungkhoaitay about this PR, will be working with him to make some minor changes.

File: test_node_env/expected.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+["ptr",108]
```
Comment (samuelfangjw) [inline test_node_env/expected.txt:1]: This file shouldn't be committed to git

File: test_node_env/expected_writer.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Script } from 'vm';
+import { createWriteStream } from 'fs';
+
+let writer2 = createWriteStream("test_node_env/expected.txt");
+const args = process.argv.slice(2);
+const expected = args[0];
+
+writer2.once('open', function(fd) {
+    let r = e.runInNewContext();
```
Comment (samuelfangjw) [inline test_node_env/expected_writer.js:9]: Is there a reason variable `e` is declared after the writer is opened? Since the writer depends on variable `e`, it is possible that `e` is undefined when called here.

File: test_node_env/expected_writer.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    if (typeof r !== 'undefined') {
+        writer2.write(JSON.stringify(r));
+    } else {
+        writer2.write("undefined");
+    }
+    writer2.end();
+});
+
+let e = new Script(expected);
```
Comment (samuelfangjw) [inline test_node_env/expected_writer.js:18]: Let's use a more descriptive variable name here. How about expectedScript?

File: test_node_env/expected_writer.js (hunk 2 above)
Comment (samuelfangjw) [inline test_node_env/expected_writer.js:9]: Is there a reason we need to process this? The expected output seems to be the same as r when I run it locally.

File: test_node_env/package.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "test_node_env",
+  "version": "1.0.0",
+  "description": "A module helps to create testing enviroment",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
```
Comment (samuelfangjw) [inline test_node_env/package.json:9]: best to put the author name here, see the outer `package.json`.

File: package.json [hunk 5]
```diff
@@ -47,7 +47,10 @@
     "split": "yarn process web split",
     "svgpdf": "./scripts/svg_to_pdf.sh",
     "test": "node ./scripts/test.js",
-    "try": "cd docs_out; http-server --cors --port 8080"
+    "try": "cd docs_out; http-server --cors --port 8080",
+    "via": "yarn do via",
+    "web": "yarn process web",
```
Comment (samuelfangjw) [inline package.json:52]: Are these changes related to your PR?

**Review by samuelfangjw — CHANGES_REQUESTED** [review]
Hi @Hungkhoaitay, thanks for the effort. I've updated your branch with the following changes.

1. Formatted `nodetest.sh`. Remember to use indentation make your programs more readable.
2. Added parameters to `expected_writer.js` and `result_writer.js`. The old method of deleting the last line and writing another new last line is slow and possibly error prone.

I've added some comments below, do take a look.

File: package.json (hunk 5 above)
Comment (martin-henz) [reply package.json:52]: These lines are likely to come from a manual merge; some version issue on Hung's side: I have removed these lines some time ago and don't want them back :-)

Comment (Hungkhoaitay) [reply package.json:52]: Yes prof, it is probably my fault :-(

File: test_node_env/expected_writer.js (hunk 2 above)
Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:9]: The reason why variable `e` is declared after the writer is opened is that I just want to make changes at the last line. It still works fine so I don't think that e can be undefined.

File: test_node_env/expected_writer.js (hunk 3 above)
Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:18]: Yea, it makes sense

File: test_node_env/expected_writer.js (hunk 2 above)
Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:9]: I tried but there is the problem that I don't remember. I think that it better off having a same form when print out.

File: test_node_env/package.json (hunk 4 above)
Comment (Hungkhoaitay) [reply test_node_env/package.json]: Okee

File: test_node_env/expected_writer.js (hunk 2 above)
Comment (samuelfangjw) [reply test_node_env/expected_writer.js:9]: Could you check again? No point processing it if we don't need to. If not needed, we can remove `expected_writer` all together.

Comment (samuelfangjw) [reply test_node_env/expected_writer.js:9]: > The reason why variable `e` is declared after the writer is opened is that I just want to make changes at the last line. It still works fine so I don't think that `e` can be undefined.

Not true at all. It works on your computer, but this doesn't mean it will still work on a faster computer, or computers with a single core which are unable to open the writer and assign the variable in parallel. 

A simple example to proof this, we can simply add a delay before declaring the variable.
```diff
writer.once("open", function (fd) {
+  console.log("Function called", s);
  let r = s.runInThisContext();
  if (typeof r !== "undefined") {
    writer.write(JSON.stringify(r));
  } else {
    writer.write("undefined");
  }
  writer.end();
});

+ const delayTime = 10;
+ const delay = ms => new Promise(resolve => setTimeout(resolve, ms))
+ await delay(delayTime).then(console.log(`${delayTime}ms delay over`, "\n"));

let s = new Script(readFileSync(programPath));
```

On my local machine, when `delayTime` is >= 5ms, s is undefined and throws an error.

What's actually happening here:
You are relying on the assumption that 
1) opening the writer and assigning the variable is done in parallel.
2) opening the writer takes longer than defining the variable.

However, this assumption does not always hold. Always best to declare variables before you use them.

Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:9]: Oh I understand, would it be oke if I change the code to this?
Screenshot 2021-07-30 at 11 32 31 AM

File: test_node_env/expected.txt (hunk 1 above)
Comment (Hungkhoaitay) [reply test_node_env/expected.txt:1]: Fixed!

File: test_node_env/expected_writer.js (hunk 2 above)
Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:9]: Yes, when I change to this:
Screenshot 2021-07-30 at 11 39 03 AM

This is the error
Screenshot 2021-07-30 at 11 39 30 AM

Thus, I think it better off having the same writer

Comment (martin-henz) [reply test_node_env/expected_writer.js:9]: If you want to add a newline to the files, you can always add:
echo "\n" >> test_node_env/expect.txt

Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:10]: Yes prof, I tried and it works for creating a new line. However, this error appears.
Screenshot 2021-07-30 at 11 56 20 AM

Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:10]: I remember that you and me have discussed about it once

Comment (martin-henz) [reply test_node_env/expected_writer.js:10]: yes I remember we solved this by applying JSON.stringify to one of the strings.

Comment (martin-henz) [reply test_node_env/expected_writer.js:10]: or eval. Whatever works.

Comment (samuelfangjw) [reply test_node_env/expected_writer.js:10]: Yes this would work!

Comment (Hungkhoaitay) [reply test_node_env/expected_writer.js:10]: Thank you, I fixed it.

Comment (samuelfangjw) [reply test_node_env/expected_writer.js:10]: ok let's use eval then.

**Review by samuelfangjw — APPROVED** [review]
LGTM, few failing tests but doesn't seem to be related to your work.
